### PR TITLE
api: pass Subchannel state updates to SubchannelStateListener rather than LoadBalancer (take 2)

### PR DESCRIPTION
**Pull request by zhangkun83**
This is a revised version of #5503 (62b03fd7e6936b7b1e78e4f13e9944d5a6da0771), which was rolled back in f8d0868e1d0bf7bd442c62c1dc260ff682448d3b. The newer version passes SubchannelStateListener to Subchannel.start() instead of SubchannelCreationArgs, which allows us to remove the Subchannel argument from the listener, which works as a solution for #5676.

LoadBalancers that call the old `createSubchannel()` will get `start()` implicitly called with a listener that passes updates to the deprecated `LoadBalancer.handleSubchannelState()`. Those who call the new `createSubchannel()` will have to call `start()` explicitly.

This PR has 3 commits:
1. Roll forward the original change.
2. Revert GRPCLB back to the old API. See commit message for the reason.
3. The revision.

Nothing new is introduced in commits 1 and 2. Reviewers should only need to review commit 3.

### CODE REVIEW

**Comment by zhangkun83** [comment]
There is a caveat, is that if the parent LB and the child LB are using different versions of APIs, the parent won't be able to intercept Subchannel states, like what HealthCheckingLoadBalancerFactory is doing.
1. Parent LB is on the old API and the child is on the new: child passes listener to `Subchannel.start()`, which the parent is not intercepting, because the parent is intercepting `LoadBalancer.handleSubchannelState()`. Subchannel state updates go to the child listener intact.
2. Parent LB is on the new API and the child is on the old: child calls the old `createSubchannel()` which the parent is not intercepting, because the parent is intercepting the new `createSubchannel()` and returning a wrapped `Subchannel` that will intercept `start()`. ChannelImpl converts the old `createSubchannel()` into a new one followed by an implicit `start()`, which calls `handleSubchannelState()` on the parent which is unintercepted. Subchannel state updates go to the child listener intact.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 1]
```diff
@@ -1529,6 +1550,18 @@ public void run() {
 
     @Override
     public void requestConnection() {
+      syncContext.execute(new Runnable() {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Why was this changed? It seems it should be safe to check `started` here without going onto the syncContext.

File: core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java [hunk 2]
```diff
@@ -60,11 +60,11 @@
  * A {@link LoadBalancer} that provides round-robin load-balancing over the {@link
  * EquivalentAddressGroup}s from the {@link NameResolver}.
  */
-final class RoundRobinLoadBalancer extends LoadBalancer implements SubchannelStateListener {
+final class RoundRobinLoadBalancer extends LoadBalancer {
   @VisibleForTesting
   static final Attributes.Key<Ref<ConnectivityStateInfo>> STATE_INFO =
       Attributes.Key.create("state-info");
-  // package-private to avoid synthetic access
+  // package-private to avoid synthetic accessnnnnnnnn
```
Comment (ejona86) [inline core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java]: Remove a few n's :smile:

File: services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java [hunk 3]
```diff
@@ -274,13 +279,16 @@ void setServiceName(@Nullable String newServiceName) {
       adjustHealthCheck();
     }
 
-    void updateRawState(ConnectivityStateInfo rawState) {
+    public void onSubchannelState(ConnectivityStateInfo rawState) {
```
Comment (voidzcy) [inline services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java]: Isn't `HealthCheckState` implementing `SubchannelStateListener` any more?

File: api/src/main/java/io/grpc/LoadBalancer.java [hunk 4]
```diff
@@ -1164,6 +1148,7 @@ public ChannelLogger getChannelLogger() {
 
     /**
      * Asks the Subchannel to create a connection (aka transport), if there isn't an active one.
+     * Has no effect if {@link #start} has not been called or {@link #shutdown} has been called.
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply api/src/main/java/io/grpc/LoadBalancer.java]: I just realized we can't make the sync-context requirement for this method, because pick-first calls it from the data-path. I have sent out #5736

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 5]
```diff
@@ -1475,41 +1507,55 @@ ClientTransport obtainActiveTransport() {
...
-            return;
+      syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            internalShutdown();
           }
+        });
+    }
+
+    private void internalShutdown() {
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: This method is only used in one place. Consider move it to a named inner class?

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 6]
```diff
@@ -1475,41 +1507,55 @@ ClientTransport obtainActiveTransport() {
...
+      syncContext.throwIfNotInThisSynchronizationContext();
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: It's obvious in syncContext already.

File: api/src/test/java/io/grpc/ForwardingTestUtil.java [hunk 7]
```diff
@@ -97,7 +97,7 @@ public Object get(Method method, int argPos, Class<?> clazz) {
       try {
         method.invoke(verify(mockDelegate), args);
       } catch (InvocationTargetException e) {
-        throw new AssertionError(String.format("Method was not forwarded: %s", method));
+        throw new RuntimeException(String.format("Method was not forwarded: %s", method), e);
```
[the remark this answers is not in the document]
Comment (ejona86) [reply api/src/test/java/io/grpc/ForwardingTestUtil.java]: There's no need to subclass AssertionError. If we can't use the two-arg constructor because of our Android restrictions, then:

```java
throw new AssertionError(String.format("Method was not forwarded: %s", method)).initCause(e);
```

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 8]
```diff
@@ -1038,103 +1047,47 @@ void remove(RetriableStream<?> retriableStream) {
...
-      }
+    @Override
+    public AbstractSubchannel createSubchannel(CreateSubchannelArgs args) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      return createSubchannelInternal(args);
+    }
 
-      syncContext.execute(new AddSubchannel());
-      return subchannel;
+    private AbstractSubchannel createSubchannelInternal(final CreateSubchannelArgs args) {
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: nit: `final` not needed.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 9]
```diff
@@ -1448,18 +1401,97 @@ private void handleErrorInSyncContext(Status error) {
...
+      syncContext.throwIfNotInThisSynchronizationContext();
+      checkState(!started, "already started");
+      checkState(!shutdown, "already shutdown");
+      started = true;
+      this.listener = listener;
+      if (terminating) {
+        syncContext.execute(new Runnable() {
+            @Override
+            public void run() {
+              listener.onSubchannelState(ConnectivityStateInfo.forNonError(SHUTDOWN));
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: The javadoc says `SHUTDOWN can be safely ignored`, do we need call `onSubchannelState(SHUTDOWN)`?

File: api/src/main/java/io/grpc/LoadBalancer.java [hunk 10]
```diff
@@ -1154,6 +1124,20 @@ public ChannelLogger getChannelLogger() {
   @ThreadSafe
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1771")
   public abstract static class Subchannel {
+    /**
+     * Starts the Subchannel.  Can only be called once.
+     *
+     * <p>This method <strong>must be called from the {@link #getSynchronizationContext
+     * Synchronization Context}</strong>, otherwise it may throw. This is to avoid the race between
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply api/src/main/java/io/grpc/LoadBalancer.java]: Rephrased.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 1 above)
Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Reverted per our discussion. There is a happens-after relationship between `start()` and the other methods, thus `started` don't need to be read in the syncContext.

File: api/src/main/java/io/grpc/LoadBalancer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhangkun83) [reply api/src/main/java/io/grpc/LoadBalancer.java]: Per our discussion offline, we will keep the sync-context requirement, and all methods except `shutdown()` should be called after `start()`.

File: services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java (hunk 3 above)
Comment (zhangkun83) [reply services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java]: You are right. Made `HealthCheckState` implement `SubchannelStateListener` again.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 11]
```diff
@@ -881,6 +883,13 @@ private void maybeTerminateChannel() {
     }
   }
 
+  // Must be called from syncContext
+  private void handleInternalSubchannelState(ConnectivityStateInfo newState) {
```
[the remark this answers is not in the document]
Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: This method doesn't need to be on `LbHelperImpl` in the first place. What it does is in the scope of the top-level class.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 5 above)
Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: This method is intended to replace `shutdown()` when we are ready to turn the warning into an exception. I have added a comment to make that clear.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 6 above)
Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Please see my response above.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 9 above)
Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: The paragraph where this statement is from is the continuation of the previous paragraph, which talks about the need to create a picker for each update. Although a LoadBalancer usually don't need to react to a SHUTDOWN update, the ChannelImpl still need to propagate it for completeness. And in some cases e.g., HealthCheckingLoadBalancer and CachedSubchannelHelper, the SHUTDOWN update is actually consumed for cleanups. I have rephrased the javadoc to prevent confusion.

**Comment by zhangkun83** [comment]
@ejona86 @dapengzhang0 @voidzcy thanks for the review. PTAL.
